src/rust/engine/src/externs/mod.rs

```diff
@@ -340,16 +388,33 @@ pub fn call_function<T: AsRef<PyObject>>(func: T, args: &[Value]) -> Result<PyOb
   func.call(gil.python(), args_tuple, None)
 }
 
-pub fn generator_send(generator: &Value, arg: &Value) -> Result<GeneratorResponse, Failure> {
-  let response = with_externs(|py, e| {
-    e.call_method(
-      py,
-      "generator_send",
-      (generator as &PyObject, arg as &PyObject),
-      None,
-    )
-    .map_err(|py_err| Failure::from_py_err_with_gil(py, py_err))
-  })?;
+pub fn generator_send_or_throw(
+  generator: &Value,
+  arg: InvocationResult,
+) -> Result<GeneratorResponse, Failure> {
+  // Call the .send() or .throw() method of the coroutine, depending upon whether the result was a
```

Comment:
this can be a doc comment on the entire function

Reply:
Done!